src/themis/sym_enc_message.c

```diff
@@ -176,303 +177,371 @@ themis_status_t themis_sym_plain_decrypt(uint32_t alg,
-     */
-    if (message_length >= UINT32_MAX) {
-        return THEMIS_INVALID_PARAMETER;
+    res = soter_rand(iv, sizeof(iv));
+    if (res != THEMIS_SUCCESS) {
+        goto error;
     }
-    THEMIS_STATUS_CHECK(themis_sym_kdf(key,
-                                       key_length,
-                                       THEMIS_SYM_KDF_KEY_LABEL,
-                                       (uint8_t*)(&message_length),
-                                       sizeof(uint32_t),
-                                       in_context,
-                                       in_context_length,
-                                       derived_key,
-                                       derived_key_length),
-                        THEMIS_SUCCESS);
 
-    THEMIS_CHECK(soter_rand(iv, THEMIS_AUTH_SYM_IV_LENGTH) == SOTER_SUCCESS);
-
-    hdr->alg = THEMIS_AUTH_SYM_ALG;
-    hdr->iv_length = THEMIS_AUTH_SYM_IV_LENGTH;
-    hdr->auth_tag_length = THEMIS_AUTH_SYM_AUTH_TAG_LENGTH;
-    hdr->message_length = (uint32_t)message_length;
-
-    THEMIS_STATUS_CHECK(themis_auth_sym_plain_encrypt(THEMIS_AUTH_SYM_ALG,
-                                                      derived_key,
-                                                      derived_key_length,
-                                                      iv,
-                                                      THEMIS_AUTH_SYM_IV_LENGTH,
-                                                      in_context,
-                                                      in_context_length,
-                                                      message,
-                                                      message_length,
-                                                      encrypted_message,
-                                                      encrypted_message_length,
-                                                      auth_tag,
-                                                      &hdr->auth_tag_length),
-                        THEMIS_SUCCESS);
+    res = themis_auth_sym_plain_encrypt(hdr.alg,
+                                        derived_key,
+                                        derived_key_length,
+                                        hdr.iv,
+                                        hdr.iv_length,
+                                        user_context,
+                                        user_context_length,
+                                        message,
+                                        message_length,
+                                        encrypted_message,
+                                        encrypted_message_length,
+                                        &auth_tag[0],
+                                        &hdr.auth_tag_length);
+    if (res != THEMIS_SUCCESS) {
+        goto error;
+    }
 
-    return THEMIS_SUCCESS;
+    if (*auth_token_length < themis_scell_auth_token_key_size(&hdr)) {
```

Comment:
can we check do all checks before encryption? check size of buffer for token before encryption above?

Reply:
> can we check do all checks before encryption?

Yes, we probably can, but arguably we will need to check after encryption too. We do early checks without encryption when the caller code expects THEMIS_BUFFER_TOO_SMALL to be returned, using some hardcoded expected value for maximum possible buffer size. This check compares the buffer length with actual token length value returned by Soter after encryption. It may be possible to safely use a buffer of smaller size than maximum.

Reply:
I thought we can calculate final size of encrypted ciphertext before encryption. if not, it bad that may be case when user should check error value and compare with buffer_to_small more then once because it require to check it in some loop and probably re-allocate buffer. can we do something and allow to check error on buffer_to_small only once before encryption/decryption?

Reply:
We can (and do) conservatively estimate the maximum size of the encrypted message and return than with “buffer too small” error.

In order to _accurately_ tell the encrypted message size we need to execute everything up to and including `soter_sym_encrypt_update()` and `soter_sym_decrypt_final()` calls which return authoritative length of encrypted message and authentication token, based on the used algorithm, key, IV, and input. Those functions can return SOTER_BUFFER_TOO_SMALL, but they need a valid `soter_sym_ctx_t` (that is, we need to have performed KDF and generated IV to do this). In short, with current Soter API we need to basically successfully perform encryption to get accurate output size.

There is a way to avoid actually encrypting data but it's inconvenient to implement. We will need to duplicate code to get correct measurements of key length, IV, etc., initialize a temporary `soter_sym_ctx_t`, ask Soter what it thinks about expected output length, and return that value. Soter will effectively encrypt a zero-length message to ask cryptographic backend about the length. Though, I'm not sure this will work for decryption path.

What we do instead is give the user maximum size that Themis expects Soter to generate (because Themis has an idea what values cryptographic backend will return). This makes it highly likely that the buffer will have sufficient (and exact) size. However, it is also possible that the _actual_ length as returned by Soter will be smaller (because Themis makes a _conservative_ estimate). It is also possible that Soter will return “buffer too small” if Themis returns an incorrect maximum buffer size and the user believes it.

This is how it worked pretty much since the beginning, this refactoring does not change this estimation logic. Passphrase API behaves in the same way as it ultimately uses the same Soter API.